lib/semantics/symbol.h

```diff
@@ -169,12 +201,16 @@ class GenericDetails {
 class UnknownDetails {};
 
 using Details = std::variant<UnknownDetails, MainProgramDetails, ModuleDetails,
-      SubprogramDetails, SubprogramNameDetails, EntityDetails, UseDetails,
-      UseErrorDetails, GenericDetails>;
+      SubprogramDetails, SubprogramNameDetails, EntityDetails,
+      ObjectEntityDetails, ProcEntityDetails, UseDetails, UseErrorDetails,
+      GenericDetails>;
 std::ostream &operator<<(std::ostream &, const Details &);
 
 class Symbol {
 public:
+  ENUM_CLASS(Flag, Function, Subroutine);
```

Comment:
Will these flags always be exclusive, or do you need to be able to have both flags set at some intermediate stages of processing?  If they're exclusive, a three-way state might be clearer (Func, Subr, AsYetUnknown).

Sometimes, flags on symbols can get out of hand in a compiler; the representation can be inverted into sets of symbols and possibly be clearer.  I.e., instead of having symbols with flags, you have some sets of symbols.

Reply:
I definitely agree we shouldn't go overboard with flags. In this case I initially had isFunction() and isSubroutine() as member functions on Symbol. A set of symbols or a function mapping symbols to booleans is another choice. They all seem logically equivalent but I thought the flags approach was simplest. I think flags should be limited to things that apply to several different kinds of symbols (in this case, subprograms, generics, procedure entities).